pkg/sources/git/git.go

```diff
@@ -466,78 +483,110 @@ func (s *Git) ScanCommits(ctx context.Context, repo *git.Repository, path string
 	// get the URL metadata for reporting (may be empty)
 	urlMetadata := getSafeRemoteURL(repo, "origin")
 
-	var depth int64
+	ctx = context.WithValues(ctx, "repo", urlMetadata)
 
 	gitDir := filepath.Join(path, gitDirName)
+	meta := scanMetadata{gitDir: gitDir, urlMetadata: urlMetadata, reporter: reporter, scanOptions: scanOptions}
+	s.startWorkers(ctx, meta, commitsWorker, commitChan)
 
-	logger := ctx.Logger().WithValues("repo", urlMetadata)
-	logger.V(1).Info("scanning repo", "base", scanOptions.BaseHash, "head", scanOptions.HeadHash)
-	for commit := range commitChan {
-		if len(scanOptions.BaseHash) > 0 {
-			if commit.Hash == scanOptions.BaseHash {
-				logger.V(1).Info("reached base commit", "commit", commit.Hash)
-				break
+	return nil
+}
+
+func (s *Git) startWorkers(ctx context.Context, meta scanMetadata, worker workerFunc, commitChan <-chan gitparse.Commit) {
+	var wg sync.WaitGroup
+
+	for i := 0; i < s.concurrency; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			ctx := context.WithValue(ctx, "scan_commit_worker_id", common.RandomID(5))
+			if err := worker(ctx, s, meta, commitChan); err != nil {
+				ctx.Logger().Error(err, "error in worker")
 			}
-		}
-		if scanOptions.MaxDepth > 0 && depth >= scanOptions.MaxDepth {
-			logger.V(1).Info("reached max depth", "depth", depth)
-			break
-		}
-		depth++
-		atomic.AddUint64(&s.metrics.commitsScanned, 1)
-		logger.V(5).Info("scanning commit", "commit", commit.Hash)
-		for _, diff := range commit.Diffs {
-			if !scanOptions.Filter.Pass(diff.PathB) {
-				continue
+		}()
+	}
+
+	wg.Wait()
+}
+
+func commitsWorker(ctx context.Context, s *Git, meta scanMetadata, commitChan <-chan gitparse.Commit) error {
+	for {
+		select {
+		case commit, ok := <-commitChan:
+			if !ok {
+				return nil
 			}
 
-			fileName := diff.PathB
-			if fileName == "" {
-				continue
+			var depth int64
+
+			if len(meta.scanOptions.BaseHash) > 0 {
+				if commit.Hash == meta.scanOptions.BaseHash {
+					ctx.Logger().V(1).Info("reached base commit", "commit", commit.Hash)
+					break
```

Comment:
I *think* this breaks out of the `select` and not the `for`. Is that intentional? Maybe some labels will help here.

Reply:
Oh... actually i think we can just return since the initial implementation we broke out of the outer for and returned nil.

Reply:
I actually think we can simply return nil? Since [this break](https://github.com/trufflesecurity/trufflehog/blob/b0fab16ad4b5cc42c38ee4e2ee362aa3c47dea24/pkg/sources/git/git.go#L479C10-L479C10) would also return from the function, or is that not correct?

Reply:
Yeah, returning `nil` looks like it would preserve the original functionality.